src/riscv/lib/src/machine_state/memory/state.rs

```diff
@@ -240,6 +232,19 @@ where
         self.executable_pages
             .modify_access(address, length, perms.can_exec());
 
+        let address = address as usize;
+        let start_page = address >> super::OFFSET_BITS;
+        let end_page = address.wrapping_add(length).wrapping_sub(1) >> super::OFFSET_BITS;
```

Comment:
This calculation assumes an inclusive range, but you're constructing an exclusive one below.